File: book/geo391/hw1/tpow/objective.c
#include <rsf.h>

int main(int argc, char* argv[])
{
    int it, nt, ix, nx, ia, na;
    float *trace, *ofunc;
    float a, a0, da, t, t0, dt, s;
    sf_file in, out;

    /* initialization */
    sf_init(argc,argv);
    in = sf_input("in");
    out = sf_output("out");

    /* get trace parameters */
    if (!sf_histint(in,"n1",&nt)) sf_error("Need n1=");
    if (!sf_histfloat(in,"d1",&dt)) dt=1.;
    if (!sf_histfloat(in,"o1",&t0)) t0=0.;

    /* get number of traces */
    nx = sf_leftsize(in,1);

    if (!sf_getint("na",&na)) na=1;
    /* number of alpha values */
    if (!sf_getfloat("da",&da)) da=0.;
    /* increment in alpha */
    if (!sf_getfloat("a0",&a0)) a0=0.;
    /* first value of alpha */

    /* change output data dimensions */
    sf_putint(out,"n1",na);
    sf_putint(out,"n2",1);
    sf_putfloat(out,"d1",da);
    sf_putfloat(out,"o1",a0);

    trace = sf_floatalloc(nt);
    ofunc = sf_floatalloc(na);

    /* initialize */
    for (ia=0; ia < na; ia++) {
	ofunc[ia] = 0.;
    }

    /* loop over traces */
    for (ix=0; ix < nx; ix++) {

	/* read data */
	sf_floatread(trace,nt,in);

	/* loop over alpha */
	for (ia=0; ia < na; ia++) {
	    a = a0+ia*da;

	    /* loop over time samples */
	    for (it=0; it < nt; it++) {
		t = t0+it*dt;

		/* apply gain t^alpha */
		s = trace[it]*powf(t,a);
		
                /* !!! MODIFY THE NEXT LINE !!! */
		ofunc[ia] += s*s; 
	    }
	}
    }

    /* write output */
    sf_floatwrite(ofunc,na,out);

    exit(0);
}

File: book/geo384h/hw1/tpow/objective.py
#!/usr/bin/env python

import sys
import math
import numpy
import m8r

# initialization
par = m8r.Par()
inp = m8r.Input()
out = m8r.Output()

# get trace parameters
nt = inp.int('n1')
dt = inp.float('d1')
t0 = inp.float('o1')
 
#  get number of traces
nx = inp.leftsize(1)

na = par.int('na',1)     # number of alpha values
da = par.float('da',0.0) # increment in alpha
a0 = par.float('a0',0.0) # first value of alpha

# change output data dimensions
out.put('n1',na)
out.put('n2',1)
out.put('d1',da)
out.put('o1',a0)

trace = numpy.zeros(nt,'f')
tgain  = numpy.zeros(nt,'f')
ofunc = numpy.zeros(na,'f')

# loop over traces
for ix in range(nx):
    # read data
    inp.read(trace)

    # loop over alpha
    for ia in range(na):
        alpha = a0+ia*da

        # loop over time samples
        for it in range(nt):
            t = t0+it*dt

            # apply gain t^alpha 
            s = trace[it]*math.pow(t,alpha)
		
            # !!! MODIFY THE NEXT LINE !!! 
            ofunc[ia] += s*s

# write output
out.write(ofunc)
sys.exit(0)
